File: NostrVault/app/src/main/java/com/nostrvault/data/model/QuoteReference.kt
package com.nostrvault.data.model

import com.nostrvault.util.Bech32

/**
 * The `nostr:` references a note makes to other events, and the single string
 * form the rest of the app uses to name one.
 *
 * There is exactly one representation, because there used to be two and they
 * drifted: the parser was changed to decode bech32 to a hex event id while both
 * of its consumers still required a `note1`/`nevent1` prefix, so every lookup
 * returned null and no quoted note ever resolved. An identifier here is either
 *
 *  - a 64-character hex event id (from `note1` or `nevent1`), or
 *  - a `naddr:<kind>:<pubkey>:<d-tag>` coordinate (from `naddr1`), naming an
 *    addressable NIP-33 event that is fetched by kind/author/`d` rather than id.
 *
 * Everything here is pure string and TLV work — no relay, no Android — so it is
 * unit testable. That is also why the bech32 layer lives in [Bech32] rather than
 * in `HavenBridge`, which cannot be loaded outside a device.
 */
object QuoteReference {

    /** Marks an addressable-event coordinate rather than a 32-byte event id. */
    const val COORDINATE_PREFIX = "naddr:"

    /** `nostr:note1…` / `nostr:nevent1…` / `nostr:naddr1…` event references. */
    val REGEX = Regex(
        """nostr:(note1[a-z0-9]+|nevent1[a-z0-9]+|naddr1[a-z0-9]+)""",
        RegexOption.IGNORE_CASE,
    )

    data class Coordinate(val kind: Int, val pubkey: String, val dTag: String)

    /**
     * Every quote reference in [content], in the order they appear, decoded to
     * identifiers, with repeats and undecodable references dropped.
     *
     * Repeats are dropped because these address rows in a list: the same
     * reference twice is the same card twice, and a `key`-ed list needs its keys
     * to be unique.
     */
    fun identifiers(content: String): List<String> =
        REGEX.findAll(content)
            .mapNotNull { identifierFor(it.groupValues[1]) }
            .distinct()
            .toList()

    /** The identifier a single bech32 reference names, or null if it will not decode. */
    fun identifierFor(bech32: String): String? {
        val (hrp, payload) = Bech32.decode(bech32) ?: return null
        return when (hrp) {
            "note" -> if (payload.size == 32) payload.toHex() else null
            "nevent" -> eventIdFromNeventTLV(payload)
            "naddr" -> coordinateFromNaddrTLV(payload)
            else -> null
        }
    }

    /** True when [identifier] names an addressable event rather than an event id. */
    fun isCoordinate(identifier: String): Boolean = identifier.startsWith(COORDINATE_PREFIX)

    /** The event id carried by an `nevent` TLV payload — type 0, 32 bytes. */
    fun eventIdFromNeventTLV(payload: ByteArray): String? {
        for (entry in tlvEntries(payload)) {
            if (entry.type == 0 && entry.value.size == 32) return entry.value.toHex()
        }
        return null
    }

    /**
     * A coordinate from an `naddr` TLV payload.
     *
     * NIP-19 naddr TLV: type 0 = d-tag (UTF-8), 1 = relay, 2 = pubkey (32
     * bytes), 3 = kind (4 bytes, big endian). Kind and pubkey are both required
     * — without them the reference names no event. An empty d-tag is legal.
     */
    fun coordinateFromNaddrTLV(payload: ByteArray): String? {
        var dTag: String? = null
        var pubkey: String? = null
        var kind: Int? = null

        for (entry in tlvEntries(payload)) {
            when {
                entry.type == 0 -> dTag = String(entry.value, Charsets.UTF_8)
                entry.type == 2 && entry.value.size == 32 -> pubkey = entry.value.toHex()
                entry.type == 3 && entry.value.size == 4 ->
                    kind = entry.value.fold(0) { acc, b -> (acc shl 8) or (b.toInt() and 0xFF) }
            }
        }

        val k = kind ?: return null
        val p = pubkey ?: return null
        return coordinate(k, p, dTag ?: "")
    }

    /** Builds a coordinate string. One definition, so [parseCoordinate] cannot drift from it. */
    fun coordinate(kind: Int, pubkey: String, dTag: String): String =
        "$COORDINATE_PREFIX$kind:$pubkey:$dTag"

    /**
     * Splits a coordinate back into its parts, or null for anything that is not
     * one — a plain hex event id included.
     */
    fun parseCoordinate(identifier: String): Coordinate? {
        if (!isCoordinate(identifier)) return null
        // limit 4 keeps a d-tag containing ":" intact.
        val parts = identifier.split(":", limit = 4)
        if (parts.size < 3) return null
        val kind = parts[1].toIntOrNull() ?: return null
        if (parts[2].isEmpty()) return null
        return Coordinate(kind, parts[2], if (parts.size > 3) parts[3] else "")
    }

    // ── TLV ──────────────────────────────────────────────────────────

    private data class TLVEntry(val type: Int, val value: ByteArray)

    /**
     * Walks a NIP-19 TLV payload. A truncated entry ends the walk rather than
     * being read past — a malformed reference should yield nothing, not garbage.
     */
    private fun tlvEntries(payload: ByteArray): List<TLVEntry> {
        val entries = mutableListOf<TLVEntry>()
        var i = 0
        while (i + 2 <= payload.size) {
            val type = payload[i].toInt() and 0xFF
            val length = payload[i + 1].toInt() and 0xFF
            i += 2
            if (i + length > payload.size) break
            entries.add(TLVEntry(type, payload.copyOfRange(i, i + length)))
            i += length
        }
        return entries
    }

    private fun ByteArray.toHex(): String = joinToString("") { "%02x".format(it) }
}

File: HavenApp/HavenApp/Models/QuoteReference.swift
import Foundation

/// The `nostr:` references a note makes to other events, and the coordinate
/// string the app uses to name an addressable (NIP-33) event it has not
/// resolved yet.
///
/// The bech32 decoding itself lives in `Bech32`; everything here is pure string
/// and TLV work so it can be tested without a relay or a live app.
enum QuoteReference {
    /// Prefix marking an addressable-event coordinate rather than a 32-byte event id.
    static let coordinatePrefix = "naddr:"

    private static let referenceRegex: NSRegularExpression? = {
        try? NSRegularExpression(
            pattern: #"nostr:(note1[a-z0-9]+|nevent1[a-z0-9]+|naddr1[a-z0-9]+)"#,
            options: .caseInsensitive
        )
    }()

    /// Every `nostr:note1…` / `nostr:nevent1…` / `nostr:naddr1…` identifier in
    /// `content`, in the order they appear, with repeats dropped.
    ///
    /// Repeats are dropped because the identifiers address rows in a list: the
    /// same reference twice is the same card twice, and SwiftUI's `ForEach`
    /// needs the ids it is given to be unique.
    static func identifiers(in content: String) -> [String] {
        guard let regex = referenceRegex else { return [] }
        let ns = content as NSString
        var seen = Set<String>()
        return regex.matches(in: content, range: NSRange(location: 0, length: ns.length))
            .map { ns.substring(with: $0.range(at: 1)) }
            .filter { seen.insert($0).inserted }
    }

    /// The event id carried by an `nevent` TLV payload — type 0, 32 bytes.
    static func eventID(fromNeventTLV payload: Data) -> String? {
        for entry in tlvEntries(payload) where entry.type == 0 && entry.value.count == 32 {
            return hex(entry.value)
        }
        return nil
    }

    /// A `"naddr:<kind>:<pubkey>:<d-tag>"` coordinate from an `naddr` TLV payload.
    ///
    /// NIP-19 naddr TLV: type 0 = d-tag (UTF-8), 1 = relay, 2 = pubkey (32 bytes),
    /// 3 = kind (4 bytes, big endian). Kind and pubkey are both required — without
    /// them the reference names no event.
    static func coordinate(fromNaddrTLV payload: Data) -> String? {
        var dTag: String?
        var pubkey: String?
        var kind: UInt32?

        for entry in tlvEntries(payload) {
            switch entry.type {
            case 0: dTag = String(data: Data(entry.value), encoding: .utf8)
            case 2 where entry.value.count == 32: pubkey = hex(entry.value)
            case 3 where entry.value.count == 4:
                kind = Data(entry.value).reduce(UInt32(0)) { ($0 << 8) | UInt32($1) }
            default: break
            }
        }

        guard let kind = kind, let pubkey = pubkey else { return nil }
        return coordinate(kind: Int(kind), pubkey: pubkey, dTag: dTag ?? "")
    }

    /// Builds the coordinate string. One definition, so the parser below cannot drift from it.
    static func coordinate(kind: Int, pubkey: String, dTag: String) -> String {
        "\(coordinatePrefix)\(kind):\(pubkey):\(dTag)"
    }

    /// Splits a coordinate back into its parts. Returns nil for anything that is
    /// not a coordinate — including a plain 64-hex event id.
    static func parseCoordinate(_ coordinate: String) -> (kind: Int, pubkey: String, dTag: String)? {
        guard coordinate.hasPrefix(coordinatePrefix) else { return nil }
        // maxSplits 3 keeps a d-tag containing ":" intact.
        let parts = coordinate.split(separator: ":", maxSplits: 3, omittingEmptySubsequences: false).map(String.init)
        guard parts.count >= 3, let kind = Int(parts[1]), !parts[2].isEmpty else { return nil }
        return (kind, parts[2], parts.count > 3 ? parts[3] : "")
    }

    /// Whether events answer one quote reference.
    ///
    /// One definition for both stores: the feed keeps `FeedNote`s and the relay
    /// tab keeps `NostrEvent`s, and they must not disagree about which event a
    /// reference points at. A plain id matches by id; a coordinate matches an
    /// addressable event by kind, author and `d` tag.
    ///
    /// It is a value rather than a free function because the callers scan a list:
    /// the coordinate is parsed once here instead of once per event examined.
    struct Matcher {
        private let identifier: String
        private let wanted: (kind: Int, pubkey: String, dTag: String)?

        init(identifier: String) {
            self.identifier = identifier
            self.wanted = QuoteReference.parseCoordinate(identifier)
        }

        func matches(id: String, kind: Int, pubkey: String, tags: [[String]]) -> Bool {
            guard let wanted = wanted else { return id == identifier }
            return kind == wanted.kind && pubkey == wanted.pubkey
                && tags.contains { $0.count >= 2 && $0[0] == "d" && $0[1] == wanted.dTag }
        }
    }

    static func matcher(for identifier: String) -> Matcher { Matcher(identifier: identifier) }

    /// Single-event convenience for callers that are not scanning.
    static func event(id: String, kind: Int, pubkey: String, tags: [[String]], matches identifier: String) -> Bool {
        matcher(for: identifier).matches(id: id, kind: kind, pubkey: pubkey, tags: tags)
    }

    // MARK: - TLV

    private struct TLVEntry {
        let type: UInt8
        let value: Data
    }

    /// Walks a NIP-19 TLV payload. A truncated entry ends the walk rather than
    /// being read past — a malformed reference should yield nothing, not garbage.
    private static func tlvEntries(_ payload: Data) -> [TLVEntry] {
        var entries: [TLVEntry] = []
        var rest = Data(payload)
        while rest.count >= 2 {
            let type = rest.removeFirst()
            let length = Int(rest.removeFirst())
            guard rest.count >= length else { break }
            entries.append(TLVEntry(type: type, value: Data(rest.prefix(length))))
            rest.removeFirst(length)
        }
        return entries
    }

    private static func hex(_ data: Data) -> String {
        data.map { String(format: "%02x", $0) }.joined()
    }
}
